src/main/kotlin/org/vitrivr/cottontail/database/index/pq/PQIndex.kt

```diff
@@ -0,0 +1,380 @@
+         *
+         * @param update Collection of [DataChangeEvent]s to process.
+         */
+        override fun update(update: Collection<DataChangeEvent>) = this.withWriteLock {
+            TODO("Not yet implemented")
+        }
+
+        /**
+         * Performs a lookup through this [PQIndex.Tx] and returns a [CloseableIterator] of all [Record]s
+         * that match the [Predicate]. Only supports [KnnPredicate]s.
+         *
+         * <strong>Important:</strong> The [CloseableIterator] is not thread safe! It remains to the
+         * caller to close the [CloseableIterator]
+         *
+         * @param predicate The [Predicate] for the lookup
+         * @return The resulting [CloseableIterator]
+         */
+        override fun filter(predicate: Predicate): CloseableIterator<Record> = object : CloseableIterator<Record> {
+
+            /** Cast [AtomicBooleanPredicate] (if such a cast is possible).  */
+            private val predicate = if (predicate is KnnPredicate<*>) {
+                predicate
+            } else {
+                throw QueryException.UnsupportedPredicateException("Index '${this@PQIndex.name}' (PQ Index) does not support predicates of type '${predicate::class.simpleName}'.")
+            }
+
+            /** The [PQ] instance used for this [CloseableIterator]. */
+            private val pq = this@PQIndex.pqStore.get()
+
+            /** Prepares [PQLookupTable]s for the given query vector(s). */
+            private val lookupTables = this.predicate.query.map {
+                this.pq.getLookupTable(it, this.predicate.distance)
+            }
+
+            /** The [ArrayDeque] of [StandaloneRecord] produced by this [VAFIndex]. Evaluated lazily! */
+            private val resultsQueue: ArrayDeque<StandaloneRecord> by lazy {
+                prepareResults()
+            }
+
+            /** Flag indicating whether this [CloseableIterator] has been closed. */
+            @Volatile
+            private var closed = false
+
+            init {
+                this@Tx.withReadLock { }
+            }
+
+            override fun hasNext(): Boolean = this.resultsQueue.isNotEmpty()
+
+            override fun next(): Record = this.resultsQueue.removeFirst()
+
+            override fun close() {
+                if (!this.closed) {
+                    this.resultsQueue.clear()
+                    this.closed = true
+                }
+            }
+
+            /**
+             * Executes the kNN and prepares the results to return by this [CloseableIterator].
+             */
+            private fun prepareResults(): ArrayDeque<StandaloneRecord> {
+                /* Prepare data structures for NNS. */
+                val txn = this@Tx.context.getTx(this@PQIndex.parent) as EntityTx
+                val preKnns = Array(this.predicate.query.size) {
+                    MinHeapSelection<ComparablePair<LongArray, Double>>(this.predicate.k)
```

Comment:
I would suggest doing the "prekNN" with a k higher than the k of the query. Otherwise the effect of the "actual" knn later on is simply to re-rank and get the actual distances. However, the preKNN can be much more useful in terms of contributing to result quality if more candidates are considered here.

Of course, you could achieve a similar effect by specifying a higher k in the query than you actually need (say 1000 and then just take the top 10 results in your client application). If you use the same k for both KNNs you have the odd effect that the result set might depend dramatically on the k of your query. This effect is obviously somewhat intrinsic to any approximate knn algorithm, but if you by default do the preKNN with a higher k than what is in the query, this effect would be reduced at comparably low cost.

In any case, the above points will be most significant for low-k queries.

Reply:
Isn't the whole point of PQ that you can compare the query vector and database vectors in their quantized version and still get meaningful distances so you don't have to fetch the vectors? I've checked the reference given at the top of this class and it's not explicitly discussed there, but in the jegou paper this is an explicit advantage. Especially for Multimedia Retrieval applications, we tend to have large k and looking up the exact vectors has a significant cost.
See also the PQ implementation in ADAMpro: https://github.com/vitrivr/ADAMpro/blob/master/src/main/scala/org/vitrivr/adampro/data/index/structures/pq/PQIndex.scala

Reply:
I can confirm that Having _preK > k_ for low values of k dramatically increases the quality of the results with very little impact on execution time. Hence I currently chose preK = k * 1.15.

To Silvan's point, I guess depending on the use-cases, an argument can be made for both modes of operation. Maybe this is something that should be configurable?

Reply:
Configuration sounds great, this would also make it possible to benchmark different configuration options using something like the data from Luca’s dissertation

Reply:
Opened an issue for this.

I guess a discussion needs to be had as to how this should be implemented.